src/Orleans/CodeGeneration/GrainInterfaceData.cs

```diff
@@ -104,49 +104,49 @@ public GrainInterfaceData(Language language)
             this.language = language;
         }
 
-        public GrainInterfaceData(Language language, Type type) : this(language)
+        public GrainInterfaceData(Language language, TypeInfo typeInfo) : this(language)
         {
-            if (!IsGrainInterface(type))
-                throw new ArgumentException(String.Format("{0} is not a grain interface", type.FullName));
+            if (!IsGrainInterface(typeInfo))
+                throw new ArgumentException(String.Format("{0} is not a grain interface", typeInfo.FullName));
 
             List<string> violations;
 
-            bool ok = ValidateInterfaceRules(type, out violations);
+            bool ok = ValidateInterfaceRules(typeInfo, out violations);
 
             if (!ok && violations != null && violations.Count > 0)
-                throw new RulesViolationException(string.Format("{0} does not conform to the grain interface rules.", type.FullName), violations);
+                throw new RulesViolationException(string.Format("{0} does not conform to the grain interface rules.", typeInfo.FullName), violations);
 
-            Type = type;
+            TypeInfo = typeInfo;
             DefineClassNames(true);
         }
 
-        public static GrainInterfaceData FromGrainClass(Type grainType, Language language)
+        public static GrainInterfaceData FromGrainClass(TypeInfo grainTypeInfo, Language language)
         {
-            if (!TypeUtils.IsConcreteGrainClass(grainType) &&
-                !TypeUtils.IsSystemTargetClass(grainType))
+            if (!TypeUtils.IsConcreteGrainClass(grainTypeInfo) &&
+                !TypeUtils.IsSystemTargetClass(grainTypeInfo))
             {
-                List<string> violations = new List<string> { String.Format("{0} implements IGrain but is not a concrete Grain Class (Hint: Extend the base Grain or Grain<T> class).", grainType.FullName) };
+                List<string> violations = new List<string> { String.Format("{0} implements IGrain but is not a concrete Grain Class (Hint: Extend the base Grain or Grain<T> class).", grainTypeInfo.FullName) };
                 throw new RulesViolationException("Invalid Grain class.", violations);
             }
 
-            var gi = new GrainInterfaceData(language) { Type = grainType };
+            var gi = new GrainInterfaceData(language) { TypeInfo = grainTypeInfo };
             gi.DefineClassNames(false);
             return gi;
         }
 
-        public static bool IsGrainInterface(Type t)
+        public static bool IsGrainInterface(TypeInfo t)
         {
             if (t.IsClass)
                 return false;
-            if (t == typeof(IGrainObserver) || t == typeof(IAddressable) || t == typeof(IGrainExtension))
+            if (t.Equals(typeof(IGrainObserver)) || t.Equals(typeof(IAddressable)) || t.Equals(typeof(IGrainExtension)))
```

Comment:
Yes. I've checked. Actually, the TypeInfo does not have '==' operator in coreclr so you should use the TypeInfo.Equals method.